---
title: "Wright-Fisher model simulation"
author: "Marc-Olivier Beausoleil"
date: "`r Sys.Date()`"
output:
  html_document:
    self_contained: no
bibliography: references/wf_model_simul.bib
csl: styles/evolution.csl
---

```{r setup, include=FALSE}
knitr::opts_chunk$set(echo = TRUE, fig.align = 'center')
```

```{r prepare_R, echo = FALSE, warning=FALSE, results='hide'}
source("../scripts/0.initialize.R")

# load the functions
source("../scripts/WF_genetic_drift_simulation.R")
```

# The Wright-Fisher Model

Population genetics is a field of study interested in the precise mechanisms describing evolutionary processes [@hartl_principles_2007]. Of the 'big five' evolutionary processes (mutation, natural selection, gene flow, mating patterns, genetic drift), genetic drift can be expressed with the Wright-Fisher model which 'isolates' only the effect of genetic drift. 

Here is an example of simulation that was made in `R` to show genetic drift without natural selection. The same script, or function, can be used to generate patterns like these, but by varying the parameters (population sizes or starting alleles). 
- Each plot below were generated with 100 populations with an allele frequency p of 50% (so q = 0.5) as an initial condition. The only thing that varies from the 4 plots is the number of individuals within each population (note that the random seed is always the same). We can see now see with the simulation that 
  * No **particular** allele will be fixated (unbiased);
  * the **number of individuals** seems an important factor in the fixation of alleles. The fluctuations in allele frequency are larger in smaller populations;
  * some genetic variation can be **lost**;
  * **makes populations different**, even if the pop are identical at the beginning;
  * **Fixation** of an allele without natural selection.

```{r}
par(mfrow = c(2,2), mar = c(4,4,3,1))
nsim = 100 # number of replicate populations
ngen = 500 # number of generations 
pstart = 0.5 # Allele frequency at start 
seed = 1245
wf.simulation.n.2 = wf.drift(N = 2,   p = pstart, N.gen = ngen, N.sim = nsim, seed = seed)
wf.simulation.n.20 = wf.drift(N = 20,  p = pstart, N.gen = ngen, N.sim = nsim, seed = seed)
wf.simulation.n.200 = wf.drift(N = 200, p = pstart, N.gen = ngen, N.sim = nsim, seed = seed)
wf.simulation.n.2000 = wf.drift(N = 2000,p = pstart, N.gen = ngen, N.sim = nsim, seed = seed)
```

We can play the same game but keeping the number of individuals constant (here 20), but changing the allele frequency at the start of the simulation. 

```{r}
par(mfrow = c(2,2), mar = c(4,4,3,1))
nid = 20
wf.simulation.n.20.p.1 = wf.drift(N = nid, p = .1, N.gen = ngen, N.sim = nsim, seed = seed)
wf.simulation.n.20.p.3 = wf.drift(N = nid, p = .3, N.gen = ngen, N.sim = nsim, seed = seed)
wf.simulation.n.20.p.5 = wf.drift(N = nid, p = .5, N.gen = ngen, N.sim = nsim, seed = seed)
wf.simulation.n.20.p.9 = wf.drift(N = nid, p = .9, N.gen = ngen, N.sim = nsim, seed = seed)
```

This is the same as the previous set of plots, but the number of individuals was increased to 1000. 

```{r}
par(mfrow = c(2,2), mar = c(4,4,3,1))
nid = 1e4
wf.simulation.n.1e4.p.1 = wf.drift(N = nid, p = .1, N.gen = ngen, N.sim = nsim, seed = seed)
wf.simulation.n.1e4.p.3 = wf.drift(N = nid, p = .3, N.gen = ngen, N.sim = nsim, seed = seed)
wf.simulation.n.1e4.p.5 = wf.drift(N = nid, p = .5, N.gen = ngen, N.sim = nsim, seed = seed)
wf.simulation.n.1e4.p.9 = wf.drift(N = nid, p = .9, N.gen = ngen, N.sim = nsim, seed = seed)
```

# Hardy-Weinberg principle

Genotype frequencies is important measure in populations to understand if any evolutionary mechanism it at play [@hahn_molecular_2018]. A simple mathematical model was developed in 1908 to infer the genotype frequencies based on the allele frequency in a population [@hardy_mendelian_1908; @weinberg_uber_1908].

```{r hardy_weinber_principle, fig.width=7, fig.height=5}
par(mfrow = c(1,1), mar = c(4,4,3,1))

# Make a sequence of p allele frequencies 
p.seq = seq(1,0,by = -.01)

# Calculate expected genotype frequencies
dat.hw = t(mapply(h.p.2pq.q, p = p.seq)) 

matplot(x = p.seq, 
        y = dat.hw, 
        type = "l",
        main = paste0("Hardy-Weinberg principle"),            # Title 
        xlab = paste0("Frequency of p (A) allele (or q (a) if 1-p)"), # x-axis lab title
          ylab = "Genotype frequency",                        # y-axis lab title
        ylim = c(0,1),
        lwd = 4,
        lty = 1,
        col = c("black","blue","red"))             
legend(x = .5,
       y = 1, 
       legend = c("AA","Aa","aa"), 
       col = c("black","blue","red"), 
       lty = 1, 
       lwd = 4, 
       title = "Genotypes")
```


# Saving the data

Here I'm exporting all the simulated data generated. 

```{r export_data}
save(wf.simulation.n.2,
     wf.simulation.n.20,
     wf.simulation.n.200,
     wf.simulation.n.2000,
     wf.simulation.n.20.p.1,
     wf.simulation.n.20.p.3,
     wf.simulation.n.20.p.5,
     wf.simulation.n.20.p.9,
     wf.simulation.n.1e4.p.1,
     wf.simulation.n.1e4.p.3,
     wf.simulation.n.1e4.p.5,
     wf.simulation.n.1e4.p.9, 
     file = "../data/wf_simulation.RData")

save(dat.hw, file = "../data/hw_simulation.RData")
```


# References
